#Questions: do we have to identiy misformat in dataset that does not relato to our solution

# 1. Business Understanding

a)

- Business Problem: Optimize vehicle design to help defining sales and marketing strategies by identifying which features most influence consumer demand.
- Consider insights: The dataset including different car model and its components, which can be analyzed to predict trend, optimize price and identify which component of an automobile shoud be prioritized.

b)

Three potential data solutions:
- Train a regression to predict Popularity, then extract feature importance
  - Feasibility: high, we already have data with clear feature and target variable (Popularity)
- Train a regression mix of engine fuel type, engine HP, engine cylinders, and driven-wheels gives the best overall fuel economy (city and highway MPG)
  - Feasibility: high, already have data with clear feature and target variable (city and highway MPG)
- Using EDA and correlation analysis to identiy trend between market price and automobile
  - Feasibility: high, the dataset already includes MSRP (target) and relevant features (Engine HP, Market Category, Vehicle Size, etc.).

After careful consideration, I decided to go with second option because regression in second solution has fewer key variables, which lead to more stable prediction. The second option also tackle a specific business selling point, which for those automobile brands who want to prioritize MPG.

c) 
Hierachal domain graph


d)

| Feature Name      | Domain Concept     | Feature Description                                                                 | Feature Type | Data Type     |
|-------------------|--------------------|-------------------------------------------------------------------------------------|--------------|---------------|
| city mpg          | Fuel Economy       | Fuel efficiency in city driving conditions                                         | Target       | Numeric (int) |
| highway MPG       | Fuel Economy       | Fuel efficiency in highway driving conditions                                      | Target       | Numeric (int) |
| city highway MPG  | Fuel Economy       | Ratio of fuel efficiency while driving in city and highway, with ratio 55% city and 45% highway | Derived      | Numeric (float) |
| Engine HP         | Engine             | Engine horsepower                                                                  | Predictor    | Numeric (int) |
| Engine Cylinders  | Engine             | Number of cylinders                                                                | Predictor    | Numeric (int) |
| Engine Fuel Type  | Engine             | Type of fuel engine uses                                                           | Predictor    | Categorical   |
| Driven_Wheels     | Drivetrain Layout  | Wheels of the vehicle receiving power from the engine to propel the vehicle        | Predictor    | Categorical   |


# 2. Data Exploration

In [ ]:
import pandas as pd

df = pd.read_csv('data.csv')

Missing City MPG: 0.0%
Missing Highway MPG: 0.0%
Missing Engine Fuel Type: 0.03%
Missing Engine HP: 0.58%
Missing Engine Cylinders: 0.25%

Duplicate rows: 715

Make                  object
Model                 object
Year                   int64
Engine Fuel Type      object
Engine HP            float64
Engine Cylinders     float64
Transmission Type     object
Driven_Wheels         object
Number of Doors      float64
Market Category       object
Vehicle Size          object
Vehicle Style         object
highway MPG            int64
city mpg               int64
Popularity             int64
MSRP                   int64
dtype: object


In [54]:
# Completeness
num_rows = len(df)

num_rows_with_missing_values = df.isnull().any(axis=1).sum()

num_missing_city_mpg = df['city mpg'].isnull().sum()
num_missing_highway_mpg = df['highway MPG'].isnull().sum()
num_missing_fuel_type = df['Engine Fuel Type'].isnull().sum()
num_missing_engine_hp = df['Engine HP'].isnull().sum()
num_missing_engine_cylinders = df['Engine Cylinders'].isnull().sum()
num_missing_drive_type = df['Driven_Wheels'].isnull().sum()

print(f"Number of rows with missing values: {num_rows_with_missing_values} ({num_rows_with_missing_values / num_rows * 100:.2f}%)")

print(f"Number of missing city mpg values: {num_missing_city_mpg} ({num_missing_city_mpg / num_rows * 100:.2f}%)")
print(f"Number of missing highway mpg values: {num_missing_highway_mpg} ({num_missing_highway_mpg / num_rows * 100:.2f}%)")
print(f"Number of missing fuel type values: {num_missing_fuel_type} ({num_missing_fuel_type / num_rows * 100:.2f}%)")
print(f"Number of missing engine hp values: {num_missing_engine_hp} ({num_missing_engine_hp / num_rows * 100:.2f}%)")
print(f"Number of missing engine cylinders values: {num_missing_engine_cylinders} ({num_missing_engine_cylinders / num_rows * 100:.2f}%)")
print(f"Number of missing drive type values: {num_missing_drive_type} ({num_missing_drive_type / num_rows * 100:.2f}%)")



Number of rows with missing values: 3830 (32.15%)
Number of missing city mpg values: 0 (0.00%)
Number of missing highway mpg values: 0 (0.00%)
Number of missing fuel type values: 3 (0.03%)
Number of missing engine hp values: 69 (0.58%)
Number of missing engine cylinders values: 30 (0.25%)
Number of missing drive type values: 0 (0.00%)


In [53]:
# Uniqueness
print(f"Duplicate rows: {df.duplicated().sum()} ({df.duplicated().sum() / num_rows * 100:.2f}%)")

Duplicate rows: 715 (6.00%)


In [16]:
# Consistency
print(df.dtypes)

Make                  object
Model                 object
Year                   int64
Engine Fuel Type      object
Engine HP            float64
Engine Cylinders     float64
Transmission Type     object
Driven_Wheels         object
Number of Doors      float64
Market Category       object
Vehicle Size          object
Vehicle Style         object
highway MPG            int64
city mpg               int64
Popularity             int64
MSRP                   int64
dtype: object


In [ ]:

# Engine cylinders needs to be an integer and non-negative
negative_cylinders = (df['Engine Cylinders'] < 0).sum()
non_integer_cylinders = (df['Engine Cylinders'] % 1 != 0).sum()
print(f"Negative Engine Cylinders: {negative_cylinders} rows")
print(f"Non-integer Engine Cylinders: {non_integer_cylinders} rows")

# Engine HP, city MPG, and highway MPG should be non-negative
misformat_engine_hp = (df['Engine HP'] < 0).sum()
misformat_city_mpg = (df['city mpg'] < 0).sum()
misformat_highway_mpg = (df['highway MPG'] < 0).sum()
print(f"Negative Engine HP: {misformat_engine_hp} rows")
print(f"Negative City MPG: {misformat_city_mpg} rows")
print(f"Negative Highway MPG: {misformat_highway_mpg} rows")


# Validating categorical variables
print(f"Unique Engine Fuel values: {df['Engine Fuel Type'].unique()}")
print(f"Unique Driven Wheels values: {df['Driven_Wheels'].unique()}")


Negative Engine Cylinders: 0 rows
Non-integer Engine Cylinders: 30 rows
Negative Engine HP: 0 rows
Negative City MPG: 0 rows
Negative Highway MPG: 0 rows
Unique Engine Fuel values: ['premium unleaded (required)' 'regular unleaded'
 'premium unleaded (recommended)' 'flex-fuel (unleaded/E85)' 'diesel'
 'electric' 'flex-fuel (premium unleaded recommended/E85)' 'natural gas'
 'flex-fuel (premium unleaded required/E85)'
 'flex-fuel (unleaded/natural gas)' nan]
Unique Driven Wheels values: ['rear wheel drive' 'front wheel drive' 'all wheel drive'
 'four wheel drive']


Result findings:
- Duplicated rows: 715 (6%)
- Rows with at least 1 missing values: 